lxd/device/gpu.go

```diff
@@ -90,13 +94,76 @@ func (d *gpu) validateEnvironment() error {
 	return nil
 }
 
+func (d *gpu) createVirtualGPU() error {
+	gpus, err := resources.GetGPU()
+	if err != nil {
+		return err
+	}
+
+	for _, gpu := range gpus.Cards {
+		// Skip any cards that don't match the vendorid, pci or productid settings (if specified).
+		if (d.config["vendorid"] != "" && gpu.VendorID != d.config["vendorid"]) ||
+			(d.config["pci"] != "" && gpu.PCIAddress != d.config["pci"]) ||
+			(d.config["productid"] != "" && gpu.ProductID != d.config["productid"]) {
+			continue
+		}
+
+		foundMdev := false
+
+		for mdev := range gpu.Mdev {
+			if d.config["mdev"] == mdev {
+				foundMdev = true
+				break
+			}
+		}
+
+		if !foundMdev {
+			return fmt.Errorf("Invalid mdev %q", d.config["mdev"])
+		}
+
+		// Check if the vgpu exists before creating it.
+		v := d.volatileGet()
+
+		if v["vgpu.uuid"] != "" && shared.PathExists(fmt.Sprintf("/sys/bus/pci/devices/%s/%s", gpu.PCIAddress, v["vgpu.uuid"])) {
+			return nil
+		}
+
+		// Create the virtual gpu
+		devUUID, err := uuid.NewUUID()
+		if err != nil {
+			return errors.Wrap(err, "Failed to generate UUID")
+		}
+
+		err = ioutil.WriteFile(filepath.Join(fmt.Sprintf("/sys/bus/pci/devices/%s/mdev_supported_types/%s/create", gpu.PCIAddress, d.config["mdev"])), []byte(devUUID.String()), 200)
+		if err != nil {
```

Comment:
This should special case ErrNotExist and report that the requested profile doesn't exist.